cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java

```diff
@@ -96,4 +95,48 @@ public void getFlowsByStream(HttpRequest request, HttpResponder responder,
     programListByDataAccess(request, responder, store, dsFramework, ProgramType.FLOW, Data.STREAM,
                             namespaceId, streamId);
   }
+
+  /**
+   * Returns a list of dataset associated with namespace.
+   */
+  @GET
+  @Path("/datasets")
+  public void getDatasets(HttpRequest request, HttpResponder responder,
+                          @PathParam("namespace-id") String namespaceId) {
+    dataList(request, responder, store, dsFramework, Data.DATASET, namespaceId, null, null);
+  }
+
+  /**
+   * Returns a dataset associated with namespace.
+   */
+  @GET
+  @Path("/datasets/{dataset-id}")
+  public void getDatasetSpecification(HttpRequest request, HttpResponder responder,
+                                      @PathParam("namespace-id") String namespaceId,
+                                      @PathParam("dataset-id") final String datasetId) {
+    dataList(request, responder, store, dsFramework, Data.DATASET, namespaceId, datasetId, null);
+  }
+
+  /**
+   * Returns a list of dataset associated with application.
+   */
+  @GET
+  @Path("/apps/{app-id}/datasets")
+  public void getDatasetsByApp(HttpRequest request, HttpResponder responder,
+                               @PathParam("namespace-id") String namespaceId,
+                               @PathParam("app-id") final String appId) {
+    dataList(request, responder, store, dsFramework, Data.DATASET, namespaceId, null, appId);
+  }
+
+  /**
+   * Returns all flows associated with a dataset.
+   */
+  @GET
+  @Path("/datasets/{dataset-id}/flows")
```

Comment:
Per https://issues.cask.co/browse/CDAP-1461, is it possible to add other program types like ``/datasets/{dataset-id}/services`` and ``/datasets/{dataset-id}/mapreduce`` easily?